File: internal/output/output.go
// Project:   macbash
// File:      internal/output/output.go
// Purpose:   Format and display scan results
// Language:  Go
//
// License:   Apache-2.0
// Copyright: (c) 2025 HyperSec Pty Ltd

package output

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"strings"

	"github.com/hypersec-io/macbash/internal/fixer"
	"github.com/hypersec-io/macbash/internal/rules"
)

type Formatter struct {
	writer    io.Writer
	useColors bool
}

const (
	colorReset  = "\033[0m"
	colorRed    = "\033[31m"
	colorYellow = "\033[33m"
	colorBlue   = "\033[34m"
	colorGray   = "\033[90m"
	colorBold   = "\033[1m"
)

func New(w io.Writer) *Formatter {
	useColors := false
	if f, ok := w.(*os.File); ok {
		stat, _ := f.Stat()
		useColors = (stat.Mode() & os.ModeCharDevice) != 0
	}

	return &Formatter{
		writer:    w,
		useColors: useColors,
	}
}

func (f *Formatter) Text(matches []rules.Match) {
	if len(matches) == 0 {
		fmt.Fprintln(f.writer, f.color(colorBlue, "No issues found."))
		return
	}

	grouped := groupByFile(matches)

	for file, fileMatches := range grouped {
		fmt.Fprintf(f.writer, "\n%s\n", f.color(colorBold, file))

		for _, m := range fileMatches {
			severityColor := f.severityColor(m.Rule.Severity)
			severityStr := strings.ToUpper(string(m.Rule.Severity))

			fmt.Fprintf(f.writer, "  %s:%d:%d: %s [%s]\n",
				f.color(colorGray, fmt.Sprintf("%d", m.Line)),
				m.Line,
				m.Column,
				f.color(severityColor, severityStr),
				m.Rule.ID,
			)

			fmt.Fprintf(f.writer, "    %s\n", m.Rule.Name)
			fmt.Fprintf(f.writer, "    %s%s\n",
				f.color(colorGray, "│ "),
				highlightMatch(m.Content, m.MatchedStr, f.useColors),
			)

			if m.Rule.FixTemplate != "" {
				canAutoFix := m.Rule.FixType == rules.FixReplace
				if m.Rule.FixType == rules.FixTransform {
					canAutoFix = canTransformMatch(&m)
				}

				fixType := "Fix"
				if !canAutoFix {
					fixType = "Suggest"
				}
				fmt.Fprintf(f.writer, "    %s%s: %s\n",
					f.color(colorGray, "└─ "),
					fixType,
					m.Rule.FixTemplate,
				)

				if !canAutoFix && m.Rule.WhyUnfixable != "" {
					fmt.Fprintf(f.writer, "    %sWhy: %s\n",
						f.color(colorGray, "   "),
						f.color(colorYellow, m.Rule.WhyUnfixable),
					)
				}
			}
		}
	}

	fmt.Fprintln(f.writer)
	errors, warnings, infos := countBySeverity(matches)

	parts := []string{}
	if errors > 0 {
		parts = append(parts, f.color(colorRed, fmt.Sprintf("%d error(s)", errors)))
	}
	if warnings > 0 {
		parts = append(parts, f.color(colorYellow, fmt.Sprintf("%d warning(s)", warnings)))
	}
	if infos > 0 {
		parts = append(parts, f.color(colorBlue, fmt.Sprintf("%d info(s)", infos)))
	}

	fmt.Fprintf(f.writer, "Found %s in %d file(s)\n",
		strings.Join(parts, ", "),
		len(grouped),
	)
}

func (f *Formatter) JSON(matches []rules.Match) error {
	type jsonMatch struct {
		File         string `json:"file"`
		Line         int    `json:"line"`
		Column       int    `json:"column"`
		RuleID       string `json:"rule_id"`
		RuleName     string `json:"rule_name"`
		Severity     string `json:"severity"`
		Description  string `json:"description"`
		Content      string `json:"content"`
		MatchedStr   string `json:"matched"`
		Fix          string `json:"fix,omitempty"`
		FixType      string `json:"fix_type"`
		WhyUnfixable string `json:"why_unfixable,omitempty"`
	}

	type jsonOutput struct {
		TotalIssues int         `json:"total_issues"`
		Errors      int         `json:"errors"`
		Warnings    int         `json:"warnings"`
		Infos       int         `json:"infos"`
		Matches     []jsonMatch `json:"matches"`
	}

	errors, warnings, infos := countBySeverity(matches)

	output := jsonOutput{
		TotalIssues: len(matches),
		Errors:      errors,
		Warnings:    warnings,
		Infos:       infos,
		Matches:     make([]jsonMatch, len(matches)),
	}

	for i, m := range matches {
		output.Matches[i] = jsonMatch{
			File:         m.File,
			Line:         m.Line,
			Column:       m.Column,
			RuleID:       m.Rule.ID,
			RuleName:     m.Rule.Name,
			Severity:     string(m.Rule.Severity),
			Description:  m.Rule.Description,
			Content:      m.Content,
			MatchedStr:   m.MatchedStr,
			Fix:          m.Rule.FixTemplate,
			FixType:      string(m.Rule.FixType),
			WhyUnfixable: m.Rule.WhyUnfixable,
		}
	}

	encoder := json.NewEncoder(f.writer)
	encoder.SetIndent("", "  ")
	return encoder.Encode(output)
}

func (f *Formatter) color(code, text string) string {
	if !f.useColors {
		return text
	}
	return code + text + colorReset
}

func (f *Formatter) severityColor(s rules.Severity) string {
	switch s {
	case rules.SeverityError:
		return colorRed
	case rules.SeverityWarning:
		return colorYellow
	case rules.SeverityInfo:
		return colorBlue
	default:
		return colorReset
	}
}

func groupByFile(matches []rules.Match) map[string][]rules.Match {
	grouped := make(map[string][]rules.Match)
	for _, m := range matches {
		grouped[m.File] = append(grouped[m.File], m)
	}
	return grouped
}

func countBySeverity(matches []rules.Match) (errors, warnings, infos int) {
	for _, m := range matches {
		switch m.Rule.Severity {
		case rules.SeverityError:
			errors++
		case rules.SeverityWarning:
			warnings++
		case rules.SeverityInfo:
			infos++
		}
	}
	return
}

func highlightMatch(content, matched string, useColors bool) string {
	if !useColors || matched == "" {
		return content
	}
	return strings.Replace(content, matched, colorRed+matched+colorReset, 1)
}

func canTransformMatch(m *rules.Match) bool {
	switch m.Rule.ID {
	case "grep-perl-regex", "grep-only-matching-P":
		return fixer.CanTransformGrepP(m.Content)
	default:
		return false
	}
}

File: src/output.rs
//  Project:   macbash
//  File:      src/output.rs
//  Purpose:   Format diagnostic output as colour text or JSON
//  Language:  Rust
//
//  License:   Apache-2.0
//  Copyright: (c) 2025-2026 HYPERI PTY LIMITED

use std::io::Write;

use anstyle::{AnsiColor, Color, Style};
use serde::Serialize;
use thiserror::Error;

use crate::fixer::can_transform_grep_p;
use crate::rules::{FixType, MatchHit, RuleSet, Severity};
use crate::scanner::group_by_file;

#[derive(Debug, Error)]
pub enum OutputError {
    #[error("writing output: {0}")]
    Io(#[from] std::io::Error),
    #[error("serialising json: {0}")]
    Json(#[from] serde_json::Error),
}

pub struct Formatter<'a, W: Write> {
    w: W,
    use_colour: bool,
    rules: &'a RuleSet,
}

impl<'a, W: Write> Formatter<'a, W> {
    pub fn new(w: W, use_colour: bool, rules: &'a RuleSet) -> Self {
        Self {
            w,
            use_colour,
            rules,
        }
    }

    pub fn text(&mut self, matches: &[MatchHit]) -> Result<(), OutputError> {
        if matches.is_empty() {
            self.write_styled(blue(), "No issues found.")?;
            writeln!(self.w)?;
            return Ok(());
        }
        let grouped = group_by_file(matches);
        for (file, file_matches) in &grouped {
            writeln!(self.w)?;
            self.write_styled(bold(), file)?;
            writeln!(self.w)?;
            for m in file_matches {
                let rule = self.rules.rules.iter().find(|r| r.id == m.rule_id);
                let severity = rule.map(|r| r.severity).unwrap_or(Severity::Warning);
                let severity_str = match severity {
                    Severity::Error => "ERROR",
                    Severity::Warning => "WARNING",
                    Severity::Info => "INFO",
                };
                write!(self.w, "  ")?;
                self.write_styled(gray(), &format!("{}", m.line))?;
                write!(self.w, ":{}:{}: ", m.line, m.column)?;
                self.write_styled(severity_style(severity), severity_str)?;
                writeln!(self.w, " [{}]", m.rule_id)?;
                if let Some(r) = rule {
                    writeln!(self.w, "    {}", r.name)?;
                }
                write!(self.w, "    ")?;
                self.write_styled(gray(), "│ ")?;
                let highlighted = self.highlight(&m.content, &m.matched_str);
                writeln!(self.w, "{highlighted}")?;
                if let Some(r) = rule
                    && !r.fix_template.is_empty()
                {
                    let can = match r.fix_type {
                        FixType::Replace => true,
                        FixType::Transform => can_transform_match(&m.rule_id, &m.content),
                        _ => false,
                    };
                    let label = if can { "Fix" } else { "Suggest" };
                    write!(self.w, "    ")?;
                    self.write_styled(gray(), "└─ ")?;
                    writeln!(self.w, "{label}: {}", r.fix_template)?;
                    if !can && !r.why_unfixable.is_empty() {
                        write!(self.w, "    ")?;
                        self.write_styled(gray(), "   ")?;
                        write!(self.w, "Why: ")?;
                        self.write_styled(yellow(), &r.why_unfixable)?;
                        writeln!(self.w)?;
                    }
                }
            }
        }
        writeln!(self.w)?;
        let (errs, warns, infos) = count_by_severity(matches, self.rules);
        let mut parts: Vec<String> = Vec::new();
        if errs > 0 {
            parts.push(self.render(red(), &format!("{errs} error(s)")));
        }
        if warns > 0 {
            parts.push(self.render(yellow(), &format!("{warns} warning(s)")));
        }
        if infos > 0 {
            parts.push(self.render(blue(), &format!("{infos} info(s)")));
        }
        writeln!(
            self.w,
            "Found {} in {} file(s)",
            parts.join(", "),
            grouped.len()
        )?;
        Ok(())
    }

    pub fn json(&mut self, matches: &[MatchHit]) -> Result<(), OutputError> {
        #[derive(Serialize)]
        struct JsonMatch<'a> {
            file: &'a str,
            line: usize,
            column: usize,
            rule_id: &'a str,
            rule_name: &'a str,
            severity: &'a str,
            description: &'a str,
            content: &'a str,
            #[serde(rename = "matched")]
            matched_str: &'a str,
            #[serde(skip_serializing_if = "str::is_empty")]
            fix: &'a str,
            fix_type: &'a str,
            #[serde(skip_serializing_if = "str::is_empty")]
            why_unfixable: &'a str,
        }
        #[derive(Serialize)]
        struct JsonOutput<'a> {
            total_issues: usize,
            errors: usize,
            warnings: usize,
            infos: usize,
            matches: Vec<JsonMatch<'a>>,
        }
        let (errs, warns, infos) = count_by_severity(matches, self.rules);
        let items: Vec<JsonMatch<'_>> = matches
            .iter()
            .map(|m| {
                let r = self.rules.rules.iter().find(|r| r.id == m.rule_id);
                JsonMatch {
                    file: &m.file,
                    line: m.line,
                    column: m.column,
                    rule_id: &m.rule_id,
                    rule_name: r.map(|r| r.name.as_str()).unwrap_or(""),
                    severity: r.map(|r| r.severity.as_str()).unwrap_or("warning"),
                    description: r.map(|r| r.description.as_str()).unwrap_or(""),
                    content: &m.content,
                    matched_str: &m.matched_str,
                    fix: r.map(|r| r.fix_template.as_str()).unwrap_or(""),
                    fix_type: r.map(|r| r.fix_type.as_str()).unwrap_or("suggest"),
                    why_unfixable: r.map(|r| r.why_unfixable.as_str()).unwrap_or(""),
                }
            })
            .collect();
        let out = JsonOutput {
            total_issues: matches.len(),
            errors: errs,
            warnings: warns,
            infos,
            matches: items,
        };
        let pretty = serde_json::to_string_pretty(&out)?;
        self.w.write_all(pretty.as_bytes())?;
        self.w.write_all(b"\n")?;
        Ok(())
    }

    fn write_styled(&mut self, style: Style, text: &str) -> Result<(), OutputError> {
        if self.use_colour {
            write!(
                self.w,
                "{}{text}{}",
                style.render(),
                style.render_reset()
            )?;
        } else {
            self.w.write_all(text.as_bytes())?;
        }
        Ok(())
    }

    fn render(&self, style: Style, text: &str) -> String {
        if self.use_colour {
            format!("{}{text}{}", style.render(), style.render_reset())
        } else {
            text.to_string()
        }
    }

    fn highlight(&self, content: &str, matched: &str) -> String {
        if !self.use_colour || matched.is_empty() {
            return content.to_string();
        }
        content.replacen(matched, &self.render(red(), matched), 1)
    }
}

fn count_by_severity(matches: &[MatchHit], rules: &RuleSet) -> (usize, usize, usize) {
    let (mut e, mut w, mut i) = (0_usize, 0, 0);
    for m in matches {
        let sev = rules
            .rules
            .iter()
            .find(|r| r.id == m.rule_id)
            .map(|r| r.severity);
        match sev {
            Some(Severity::Error) => e += 1,
            Some(Severity::Warning) => w += 1,
            Some(Severity::Info) => i += 1,
            None => {}
        }
    }
    (e, w, i)
}

fn can_transform_match(rule_id: &str, content: &str) -> bool {
    matches!(rule_id, "grep-perl-regex" | "grep-only-matching-P")
        && can_transform_grep_p(content)
}

fn severity_style(s: Severity) -> Style {
    match s {
        Severity::Error => red(),
        Severity::Warning => yellow(),
        Severity::Info => blue(),
    }
}

const fn red() -> Style {
    Style::new().fg_color(Some(Color::Ansi(AnsiColor::Red)))
}
const fn yellow() -> Style {
    Style::new().fg_color(Some(Color::Ansi(AnsiColor::Yellow)))
}
const fn blue() -> Style {
    Style::new().fg_color(Some(Color::Ansi(AnsiColor::Blue)))
}
const fn gray() -> Style {
    Style::new().fg_color(Some(Color::Ansi(AnsiColor::BrightBlack)))
}
const fn bold() -> Style {
    Style::new().bold()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::rules::load_builtin;
    use crate::scanner::Scanner;

    #[test]
    fn text_no_issues_says_so() {
        let rs = load_builtin().unwrap();
        let mut buf = Vec::new();
        Formatter::new(&mut buf, false, &rs).text(&[]).unwrap();
        let s = String::from_utf8(buf).unwrap();
        assert!(s.contains("No issues"));
    }

    #[test]
    fn json_schema_field_names_match_go_oracle() {
        let rs = load_builtin().unwrap();
        let s = Scanner::new(&rs).unwrap();
        let ms = s.scan_text("sed -i 's/a/b/' f\n", "t.sh");
        let mut buf = Vec::new();
        Formatter::new(&mut buf, false, &rs).json(&ms).unwrap();
        let s = String::from_utf8(buf).unwrap();
        for key in &[
            "\"total_issues\"",
            "\"errors\"",
            "\"warnings\"",
            "\"infos\"",
            "\"matches\"",
            "\"file\"",
            "\"line\"",
            "\"column\"",
            "\"rule_id\"",
            "\"rule_name\"",
            "\"severity\"",
            "\"description\"",
            "\"content\"",
            "\"matched\"",
            "\"fix_type\"",
        ] {
            assert!(s.contains(*key), "missing key {key} in:\n{s}");
        }
    }

    #[test]
    fn json_total_issues_matches_matches_array_length() {
        let rs = load_builtin().unwrap();
        let s = Scanner::new(&rs).unwrap();
        let ms = s.scan_text(
            "#!/bin/bash\nsed -i 's/a/b/' f\ngrep -P '\\d+' f\n",
            "t.sh",
        );
        let mut buf = Vec::new();
        Formatter::new(&mut buf, false, &rs).json(&ms).unwrap();
        let v: serde_json::Value = serde_json::from_slice(&buf).unwrap();
        let n = v["total_issues"].as_u64().unwrap() as usize;
        assert_eq!(n, v["matches"].as_array().unwrap().len());
        assert!(n >= 2);
    }

    #[test]
    fn text_groups_matches_under_file_headers() {
        let rs = load_builtin().unwrap();
        let s = Scanner::new(&rs).unwrap();
        let mut ms = s.scan_text("sed -i 's/a/b/' f\n", "alpha.sh");
        ms.extend(s.scan_text("grep -P '\\d+' f\n", "beta.sh"));
        let mut buf = Vec::new();
        Formatter::new(&mut buf, false, &rs).text(&ms).unwrap();
        let out = String::from_utf8(buf).unwrap();
        assert!(out.contains("alpha.sh"));
        assert!(out.contains("beta.sh"));
        assert!(out.contains("Found"));
    }

    #[test]
    fn json_severity_field_uses_lowercase_strings() {
        let rs = load_builtin().unwrap();
        let s = Scanner::new(&rs).unwrap();
        let ms = s.scan_text("sed -i 's/a/b/' f\n", "t.sh");
        let mut buf = Vec::new();
        Formatter::new(&mut buf, false, &rs).json(&ms).unwrap();
        let v: serde_json::Value = serde_json::from_slice(&buf).unwrap();
        let sev = v["matches"][0]["severity"].as_str().unwrap();
        assert!(
            sev == "error" || sev == "warning" || sev == "info",
            "unexpected severity {sev:?}"
        );
    }
}
